### PR TITLE
Fix bug where reloading dettl object was leaving stale connections mid transaction which blocked further db interaction

### PR DESCRIPTION
**Pull request by r-ash**
Note that we talked about adding a `finalize` method to the dettl object which would ensure DB connection was closed before garbage collection but this reliably causes a segfault when running the test suite. Including when I've tried to wrap the `dbDisconnect` call in a check if the connection is still valid.

### CODE REVIEW

File: R/dettl.R [hunk 1]
```diff
@@ -42,7 +44,25 @@ DataImport <- R6::R6Class(
...
     db_name = NULL,
     mode = NULL,
-    transaction = TRUE
+    load_in_transaction = TRUE,
+
+    #' @description
+    #' Tidy up open connections, rolling back any active transactions
+    close_connection = function() {
+      if (private$transaction_active) {
+        warning("Rolling back active transaction")
```
Comment (richfitz) [inline R/dettl.R]: I am not sold on warning here - warnings get grouped together and are displayed at some future point. If you *really* want a warning, make it immediate, but TBH this looks more like a message (we're telling the user we're doing something) than anything else (I generally just avoid warnings)

File: R/dettl.R [hunk 2]
```diff
@@ -42,7 +44,25 @@ DataImport <- R6::R6Class(
...
+        self$rollback_transaction()
+      }
+      ## Connection will always be null on first call to reload
+      if (!is.null(private$con)) {
+        tryCatch(
+          DBI::dbDisconnect(private$con),
+          error = function(e) {
+            ## Consume the error
```
Comment (richfitz) [inline R/dettl.R]: I think a notice here is still useful:

```
tryCatch(
  DBI::dbDisconnect(private$con),
  error = function(e) message("While disconnecting from db, ignored error:", e$message))
```

File: R/dettl.R [hunk 3]
```diff
@@ -157,11 +178,37 @@ DataImport <- R6::R6Class(
       if (!force && !dry_run && !git_repo_is_clean(self$path)) {
         stop("Can't run load as repository has unstaged changes. Update git or run in dry-run mode.")
       }
-      run_load(private$con, private$load_, private$extracted_data,
-               private$transformed_data, private$test_queries,
-               private$load_pre_, private$load_post_, self$path,
-               private$load_test_, private$transaction, dry_run,
-               private$log_table, comment)
+
+      use_transaction <- private$load_in_transaction || dry_run
```
Comment (richfitz) [inline R/dettl.R:210]: This is the only place where `load_in_transaction` is used - perhaps drop it from the class and just use `dettl_config$dettl$transaction`

File: R/dettl.R [hunk 4]
```diff
@@ -157,11 +178,37 @@ DataImport <- R6::R6Class(
...
+      if (use_transaction) {
+        self$begin_transaction()
+      }
+      message(
+        sprintf("Running load %s:",
+                use_transaction %?% "in a transaction" %:% "not in a transaction"))
```
Comment (richfitz) [inline R/dettl.R]: while this is cute, I wold have written this as:

```
      message(
        sprintf("Running load %s:",
                if (use_transaction) "in a transaction" else "not in a transaction"))
```

or even more concisely but perhaps less nicely

```
      message(
        sprintf("Running load %sin a transaction:",
                if (use_transaction) "" else "not"))
```

File: R/dettl.R [hunk 5]
```diff
@@ -19,13 +19,15 @@
 #' import$transform()
 #' import$load()
 #'
-
+# nolint start
```
Comment (richfitz) [inline R/dettl.R:22]: This seems like a bit of a flaw in lintr that we can't just disable the complexity check here, and that it is flagging up the whole class and not the methods

**Review by richfitz — CHANGES_REQUESTED** [review]

File: R/dettl.R (hunk 5 above)
Comment (r-ash) [reply R/dettl.R:22]: Yeah bit of a shame it doesn't tell us where the actual problem is - we can disable the complexity check for the whole file

File: R/dettl.R (hunk 1 above)
Comment (r-ash) [reply R/dettl.R]: Yeah that is a message really - have updated that

File: R/dettl.R (hunk 2 above)
Comment (r-ash) [reply R/dettl.R]: Fixed

File: R/dettl.R (hunk 3 above)
Comment (r-ash) [reply R/dettl.R:210]: We don't have the config at this point

File: R/dettl.R (hunk 4 above)
Comment (r-ash) [reply R/dettl.R]: Tidied that up

**Review by richfitz — APPROVED** [review]

**Pull request merged**